server/tilesource/base.py

```diff
@@ -96,25 +98,45 @@ class TileSource(object):
     cache = tileCache
     cache_lock = tileLock
 
-    def __init__(self, *args, **kwargs):
+    def __init__(self, jpegQuality=95, jpegSubsampling=0,
+                 encoding='JPEG', edge=False, *args, **kwargs):
+        """
+        Initialize the tile class.
+
+        :param jpegQuality: when serving jpegs, use this quality.
+        :param jpegSubsampling: when serving jpegs, use this subsampling (0 is
+                                full chroma, 1 is half, 2 is quarter).
+        :param encoding: 'JPEG' or 'PNG'.
+        :param edge: False to leave edge tiles whole, True or 'crop' to crop
+            edge tiles, otherwise, an #rrggbb color to fill edges.
+        """
         self.tileWidth = None
         self.tileHeight = None
         self.levels = None
         self.sizeX = None
         self.sizeY = None
 
+        if encoding not in ('PNG', 'JPEG'):
+            raise ValueError('Invalid encoding "%s"' % encoding)
+
+        self.encoding = encoding
+        self.jpegQuality = int(jpegQuality)
+        self.jpegSubsampling = int(jpegSubsampling)
+        self.edge = edge
+
         self.getThumbnail = cached(
             self.cache, key=self.wrapKey, lock=self.cache_lock)(
                 self.getThumbnail)
         self.getTile = cached(
             self.cache, key=self.wrapKey, lock=self.cache_lock)(self.getTile)
 
-    def wrapKey(self, *args, **kwargs):
-        return strhash(self.getState()) + strhash(*args, **kwargs)
-
     @abc.abstractmethod
     def getState(self):
-        return None
+        return str(self.encoding) + ',' + str(self.jpegQuality) + ',' + str(
+            self.jpegSubsampling) + ',' + str(self.edge)
```

Comment:
Is there a reason to have an implementation of a declared `abstractmethod`? Also, doesn't using `abc.abstractmethod` require you to specify `__metaclass__ = abc.ABCMeta`?